lib/improver/spotdata_new/spot_extraction.py

```diff
@@ -0,0 +1,247 @@
+        self.grid_metadata_identifier = grid_metadata_identifier
+
+    def __repr__(self):
+        """Represent the configured plugin instance as a string."""
+        return ('<SpotExtraction: neighbour_selection_method: {}, '
+                'grid_metadata_identifier: {}>'.format(
+                    self.neighbour_selection_method,
+                    self.grid_metadata_identifier))
+
+    def extract_coordinates(self, neighbour_cube):
+        """
+        Extract the desired set of grid coordinates that correspond to spot
+        sites from the neighbour cube.
+
+        Args:
+            neighbour_cube (iris.cube.Cube):
+                A cube containing information about the spot data sites and
+                their grid point neighbours.
+        Returns:
+            coordinate_cube (iris.cube.Cube):
+                A cube containing only the x and y grid coordinates for the
+                grid point neighbours given the chosen neighbour selection
+                method. The neighbour cube contains the indices stored as
+                floating point values, so they are converted to integers
+                in this cube.
+        Raises:
+            ValueError if the neighbour_selection_method expected is not found
+            in the neighbour cube.
+        """
+        method = iris.Constraint(
+            neighbour_selection_method_name=self.neighbour_selection_method)
+        index_constraint = iris.Constraint(
+            grid_attributes_key=['x_index', 'y_index'])
+        coordinate_cube = neighbour_cube.extract(method & index_constraint)
+        if coordinate_cube:
+            coordinate_cube.data = np.rint(coordinate_cube.data).astype(int)
+            return coordinate_cube
+
+        available_methods = (
+            neighbour_cube.coord('neighbour_selection_method_name').points)
+        raise ValueError(
+            'The requested neighbour_selection_method "{}" is not available in'
+            ' this neighbour_cube. Available methods are: {}.'.format(
+                self.neighbour_selection_method, available_methods))
+
+    @staticmethod
+    def extract_diagnostic_data(coordinate_cube, diagnostic_cube):
+        """
+        Extracts diagnostic data from the desired grid points in the diagnostic
+        cube.
+
+        Args:
+            coordinate_cube (iris.cube.Cube):
+                A cube containing the x and y grid coordinates for the grid
+                point neighbours.
+            diagnostic_cube (iris.cube.Cube):
+                A cube of diagnostic data from which spot data is being taken.
+        Returns:
+            spot_values (np.array):
+                An array of diagnostic values at the grid coordinates found
+                within the coordinate cube.
+        """
+        diagnostic_cube = enforce_coordinate_ordering(
+            diagnostic_cube, [diagnostic_cube.coord(axis='x').name(),
+                              diagnostic_cube.coord(axis='y').name()])
```

Comment:
Why are we using (x, y) coordinate ordering here, when we use (y, x) elsewhere in IMPROVER?

Reply:
Because some of the tools such as cartopy transform coordinates require x-y ordering. This means we either have to pass things into different functions (in neighbour finding) in different orders or we just embrace x-y throughout. The neighbour finding has already been written using x-y ordering, so this is required to match the indices that it produces.

Reply:
OK, that's fine then.  Might be worth a note in the docstring explaining this.